### DISCUSSION TITLE
macOS 如何设置开机启动 Alist？

### DISCUSSION THREAD

**Discussion by y2-Whyto** [AlistGo/alist#1759] [Q&A] [answered]
[手动安装](https://alist.nn.ci/zh/guide/install/manual.html#daemon)页面下方给出了 Linux 下守护进程的设置方式，但并不适用于 macOS，Google 上也没有搜索到 Alist 在 macOS 下配置开机启动的方法，为此我搜索了使用 launchd 配置 LaunchAgents / LaunchDaemons 的方法。

我尝试在 `~/Library/LaunchAgents` 下加入了 `alist.plist` 文件，并使用 `launchctl load -w ~/Library/LaunchAgents/alist.plist` 进行加载，但是 `launchctl list | grep alist` 中查询到 `alist` 的 `Status` 为 `1`，无 `PID`，重启后也无法访问 localhost:5244，即未能成功配置启动。

![IMAGE 2022-09-23 06:10:56](https://user-images.githubusercontent.com/42339192/191860378-0484e9be-9562-4925-b190-063fb0be2982.jpg)
![IMAGE 2022-09-23 06:11:05](https://user-images.githubusercontent.com/42339192/191860395-00052c1b-5795-4a52-9334-aaf6283d6536.jpg)

我的 `alist.plist` 文件内容为：

```xml
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>KeepAlive</key>
	<true/>
	<key>Label</key>
	<string>alist</string>
	<key>ProgramArguments</key>
	<array>
		<string>/Users/whyto/alist</string>
		<string>server</string>
	</array>
	<key>RunAtLoad</key>
	<true/>
</dict>
</plist>
```

尝试无果后，我也将其放入 `/Library/LaunchDaemons` 中利用 `sudo` 加载，但是依然没有效果。

请问如何在 macOS 下配置 Alist 开机启动呢？

**Comment by Mrered** [comment]
同问，我甚至连.plist文件都不会写，希望出一个帮助文档

**Comment by Mrered** [comment]
我实现了曲线救国

首先打开`自动操作.app`，文稿类型选择`应用程序`，添加操作`运行Shell脚本`，键入以下内容：

```
cd /Users/<你的用户名>/Applications/Alist
./alist server
```
我把`alist`放在个人程序文件夹下，你直接`cd`到你实际存放的位置。

保存该`.app`文件，再在`系统设置`-`通用`-`登录项`-`登录时打开`中添加刚刚生成的`.app`文件。

缺点是运行的时候菜单栏会有一个图标，当然你可以用诸如 [Bartender 4](https://www.macbartender.com/) 之类的软件将其隐藏掉。

**Reply by Gasiro** [reply]
### _**这个方法适用于M2 Pro芯片**_

我找到了一个办法，可以后台直接kill掉进程。菜单栏就不会出现那个一直转的图标。原理就是在这个`.app`程序外嵌套一层自动化。

比如：我在 Application 文件夹中新建一个文件夹叫 “AList” 用来存放所有自动化制作的`.app`。
我制作了三个：分别是 **① `AList`   ② `StartAList `  ③ `EndAList`**
<img width="368" alt="Icons" src="https://user-images.githubusercontent.com/117976338/222932755-f967eb76-2a95-4421-b2d8-57761539e4f8.png">

### 第一个app
其中 ① `AList` 这个`.app`里封装的命令是
`cd /Users/<你的用户名>/Applications/Alist
./alist server`

### 第二个app
② `StartAList` 这个app里封装的是几个步骤，分别是：1.开机后等待20秒   2. 启动第一个`.app` ①`AList`   3. 再等待20秒  4. 启动第三个`.app` ③ `EndAList` 
<img width="596" alt="②" src="https://user-images.githubusercontent.com/117976338/222932891-b4e41bde-9840-40b1-abb7-ef94ae8f7c71.png">

### 第三个app
③ `EndAList` 这个`.app`里封装的代码是：运行shell，找到alist的进程pid，并在在系统里直接kill掉这个进程（即kill掉菜单栏那个一直转的图标）
寻找这个进程并kill的代码如下，这个代码应该不需要输入密码：
`ps -e | grep "AList" | awk '{print $1}' | head -n1 | xargs -I {} kill -15 {}`
<img width="595" alt="Screenshot 2023-03-04 at 23 18 54" src="https://user-images.githubusercontent.com/117976338/222932980-aeae194f-c883-4c33-a497-e8737c75ab6a.png">

### 最后一步：
把第二个自动化的程序直接加入开机启动即可： 即 ② `StartAList` 

然后分享一下我自己改造的这几个图标
第一个可以在Alist官网下
第二个
![StartAList](https://user-images.githubusercontent.com/117976338/222933080-ed853133-7fc5-4274-9b41-2d4f1794f273.svg)
第三个
![EndAList](https://user-images.githubusercontent.com/117976338/222933082-c22c54bf-33ad-4ee9-8a3f-1b2635f4f6a2.svg)

**Reply by lyc59621** [reply]
兄弟，自动操作，怎么自定义图标，我的怎么一直是机器人图标

**Reply by Gasiro** [reply]
右键应用程序显示基础信息，直接把svg拖到这个位置就可以更换图标了。😄😄
![66](https://user-images.githubusercontent.com/117976338/226190361-6517770c-9850-43ac-b5f6-cccf955ef11b.jpeg)

**Comment by f63531119** [comment]
这个sh加到计划任务多简单

**Reply by y2-Whyto** [reply]
> 这个sh加到计划任务多简单

我用sh好像也不行

**Reply by f63531119** [reply]
那就你写的有问题或者权限没给

**Reply by f63531119** [reply]
#!/bin/sh

if [ -e /mnt/mmcblk2p4/alist/alist ]; then
    COUNT=$(ps | grep "./alist server" | grep -v grep | wc -l)
    echo $COUNT
    if [ $COUNT -ne 0 ]; then
        logger -t "【alist程序】" "已启动"
    else
        logger -t "【alist程序】" "未启动，尝试启动"
        cd /mnt/mmcblk2p4/alist
        nohup ./alist server > /dev/null 2>&1 &
        logger -t "【alist程序】" "启动完成"
    fi
else
    logger -t "【alist程序】" "文件不存在，停止启动操作"
fi

#用我写的注意2 10文件目录自己改 连带alist权限给了sh格式 放一起 计划任务

**Comment by y2-Whyto** [comment, marked as answer]
经过进一步探索，目前我已经解决使用 macOS 原生 launchd.plist 实现登录启动的问题了。方法：

1. 使用文本编辑器创建`local.alist.launch.plist`（以此名称为例），并拷贝至`主文件夹/资源库/LaunchAgents`下。文件内容如下：

    ```xml
    <?xml version="1.0" encoding="UTF-8"?>
    <!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
    <plist version="1.0">
    <dict>
	    <key>KeepAlive</key>
	    <dict>
		    <key>NetworkState</key>
		    <true/>
	    </dict>
	    <key>Label</key>
	    <string>local.alist.launch</string>
	    <key>ProgramArguments</key>
	    <array>
		    <string>/path/to/alist</string>
		    <string>server</string>
	    </array>
	    <key>RunAtLoad</key>
	    <true/>
	    <key>StandardErrorPath</key>
	    <string>/path/to/alist.output.log</string>
	    <key>WorkingDirectory</key>
	    <string>/path/to/alist/directory</string>
    </dict>
    </plist>
    ```

    其中：
    
    - `/path/to/alist` 是 alist 二进制程序的**绝对**路径，请自行调整。
    - `/path/to/alist.output.log` 是 alist 启动时输出的内容，存放在了名为 `alist.output.log`的文件中，可以自行设置，建议和 alist 程序放在一起。
    - `/path/to/alist/directory` 是 存放 alist 二进制程序的**目录**的**绝对**路径，即比第一个 `/path/to/alist` 少了最后的“alist”。

2. 之后在终端中输入 `launchctl load -w `（注意**末尾空格**），再将刚刚拷贝到 `LaunchAgents` 目录下的 plist 文件拖拽到终端内释放（即输入文件路径），之后按下 return 键。plist 文件会立即加载，并从下次登录起自动加载（即登录启动）。

    - 如果没有输出，则在终端中输入`ps -ef | grep alist`，应当能够看到 `/path/to/alist server` 项。此即配置成功，可以打开浏览器访问验证。（当然，前提是你之前没有手动打开 alist）
    - 如果加载时遇到报错，或者没有在 `ps -ef | grep alist` 后看到预期的输出，则是 plist 文件编写存在问题。可先使用 `plutil -lint /path/to/plist/file` 检查语法错误，再进一步排查。
    - 如果修改了有问题的 plist 后，需要重新加载，则可以先在终端中输入 `launchctl unload -w /path/to/plist/file` 解除加载，之后再使用前面的 `launchctl load -w` 加载。

------------

最后大致说明这个文件的功能，嫌麻烦也可以直接跳过：

- `KeepAlive` -- `NetworkState` -- `true`（**必需**）：在网络连接后启动，这与官方文档中的守护进程写法类似。
- `Label`--`local.alist.launch`（**必需**）：plist 的项目名字是 `local.alist.launch`，可以自行修改。
- `ProgramArguments` -- `/path/to/alist`, `server`（**必需**）：第一个位置是二进制程序**绝对**路径，第二个及之后的位置为顺序的参数。
- `RunAtLoad` -- `true`（**必需**）：加载即启动。加载和启动是两部分，不打开此项，则仅加载不启动。
- `StandardErrorPath` -- `/path/to/alist.output.log`（可选）：将运行时的输出转存至 `/path/to/alist.output.log`（**绝对**路径），可以自行修改（包括文件名）。便于 debug，跟踪运行情况。
- `WorkingDirectory` -- `/path/to/alist/directory`（**必需**）：设置 alist 的工作目录为 `/path/to/alist/directory`，即存放 alist 二进制程序的目录的**绝对**路径（**非常重要**）。如果不设置，则默认是系统根目录 `/`，在新系统中是只读的；或者如果是可写的，则 alist 会在那里自动重新写配置文件。

如果要删除可选项，则需要将<key>及其后面对应的值（string、array之类）一并删除。推荐使用 PlistEdit Pro 操作。

------------

参考资料（可自行深入探索）：

- [A launchd Tutorial](https://www.launchd.info/)
- [Apple. Creating Launch Daemons and Agents](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingLaunchdJobs.html#//apple_ref/doc/uid/TP40001762-104142)
- [Manpagez: man pages & more. man launchd.plist(5)](https://www.manpagez.com/man/5/launchd.plist/)

**Reply by Mrered** [reply]
我照做后warning了。

```
Warning: Expecting a LaunchDaemons path since the command was ran as root. Got LaunchAgents instead.
`launchctl bootstrap` is a recommended alternative.
/Library/LaunchAgents/local.alist.launchagent.plist: Path had bad ownership/permissions
Load failed: 122: Path had bad ownership/permissions
```

于是我检索查阅了一下，参照[此页](https://juejin.cn/post/7075177588259291173)解决了。

感谢你的帮助，我举一反三把 aria2 也配置成功了！

**Reply by y2-Whyto** [reply]
> 我照做后warning了。
> 
> ```
> Warning: Expecting a LaunchDaemons path since the command was ran as root. Got LaunchAgents instead.
> `launchctl bootstrap` is a recommended alternative.
> /Library/LaunchAgents/local.alist.launchagent.plist: Path had bad ownership/permissions
> Load failed: 122: Path had bad ownership/permissions
> ```
> 
> 于是我检索查阅了一下，参照[此页](https://juejin.cn/post/7075177588259291173)解决了。
> 
> 感谢你的帮助，我举一反三把 aria2 也配置成功了！

可能是你在执行 `launchctl` 的时候用了 `sudo`。用户级的登录启动项是不需要使用 `sudo` 的（即以 `root` 的身份此行，而非用户本身）。系统级的开机启动 plist 一般放在系统级路径（`/Library/LaunchDaemons`，`/Library/LaunchAgents` 等），而且所需启动的程序**拥有者**需要是 `root`，所以可能才出现了需要 `chown` 的情况。具体细节在[参考资料 2](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingLaunchdJobs.html#//apple_ref/doc/uid/TP40001762-104142) 中有提到。

<img width="1080" alt="图片" src="https://user-images.githubusercontent.com/42339192/198755094-748022fc-bc7c-46e1-bf0b-d8374bc00a6f.png">

最后，恭喜🎉🎉

**Reply by shaopak** [reply]
> > 我照做后warning了。
> > ```
> > Warning: Expecting a LaunchDaemons path since the command was ran as root. Got LaunchAgents instead.
> > `launchctl bootstrap` is a recommended alternative.
> > /Library/LaunchAgents/local.alist.launchagent.plist: Path had bad ownership/permissions
> > Load failed: 122: Path had bad ownership/permissions
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 于是我检索查阅了一下，参照[此页](https://juejin.cn/post/7075177588259291173)解决了。
> > 感谢你的帮助，我举一反三把 aria2 也配置成功了！
> 
> 可能是你在执行 `launchctl` 的时候用了 `sudo`。用户级的登录启动项是不需要使用 `sudo` 的（即以 `root` 的身份此行，而非用户本身）。系统级的开机启动 plist 一般放在系统级路径（`/Library/LaunchDaemons`，`/Library/LaunchAgents` 等），而且所需启动的程序**拥有者**需要是 `root`，所以可能才出现了需要 `chown` 的情况。具体细节在[参考资料 2](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingLaunchdJobs.html#//apple_ref/doc/uid/TP40001762-104142) 中有提到。
> 
> <img alt="图片" width="1080" src="https://user-images.githubusercontent.com/42339192/198755094-748022fc-bc7c-46e1-bf0b-d8374bc00a6f.png">
> 
> 最后，恭喜🎉🎉

执行launchctl load -w XXX后出现了如下报错，请问是什么问题
```
Load failed: 5: Input/output error
Try running `launchctl bootstrap` as root for richer errors.
```
求助

**Reply by y2-Whyto** [reply]
> > > 我照做后warning了。
> 
> > > ```
> 
> > > Warning: Expecting a LaunchDaemons path since the command was ran as root. Got LaunchAgents instead.
> 
> > > `launchctl bootstrap` is a recommended alternative.
> 
> > > /Library/LaunchAgents/local.alist.launchagent.plist: Path had bad ownership/permissions
> 
> > > Load failed: 122: Path had bad ownership/permissions
> 
> > > ```
> 
> > > 
> 
> > > 
> 
> > >     
> 
> > >       
> 
> > >     
> 
> > > 
> 
> > >       
> 
> > >     
> 
> > > 
> 
> > >     
> 
> > >   
> 
> > > 于是我检索查阅了一下，参照[此页](https://juejin.cn/post/7075177588259291173)解决了。
> 
> > > 感谢你的帮助，我举一反三把 aria2 也配置成功了！
> 
> > 
> 
> > 可能是你在执行 `launchctl` 的时候用了 `sudo`。用户级的登录启动项是不需要使用 `sudo` 的（即以 `root` 的身份此行，而非用户本身）。系统级的开机启动 plist 一般放在系统级路径（`/Library/LaunchDaemons`，`/Library/LaunchAgents` 等），而且所需启动的程序**拥有者**需要是 `root`，所以可能才出现了需要 `chown` 的情况。具体细节在[参考资料 2](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingLaunchdJobs.html#//apple_ref/doc/uid/TP40001762-104142) 中有提到。
> 
> > 
> 
> > <img alt="图片" width="1080" src="https://user-images.githubusercontent.com/42339192/198755094-748022fc-bc7c-46e1-bf0b-d8374bc00a6f.png">
> 
> > 
> 
> > 最后，恭喜🎉🎉
> 
> 
> 
> 执行launchctl load -w XXX后出现了如下报错，请问是什么问题
> 
> ```
> 
> Load failed: 5: Input/output error
> 
> Try running `launchctl bootstrap` as root for richer errors.
> 
> ```
> 
> 求助
> 
> 
请按照上面的步骤排查。更详细的讲解可参考我的博文 https://new.whyto.net/2023/01/15/launch-agent-for-alist-on-macos/

**Reply by CrazyJasonwell** [reply]
出现richer errors,肯定是plist语法问题，用plutil -lint example.plist验证一下。

**Comment by templatejacob** [comment]
是否可以用docker版的alist来实现开机启动捏

**Reply by y2-Whyto** [reply]
没有试过，可以尝试一下

**Reply by Gasiro** [reply]
docker版本好像做的很差，我试了好几遍都不行。
